### ISSUE TITLE
Possibly dangerous design in IXWebSocketMessage class

### ISSUE THREAD

**Issue by picanumber** [machinezone/IXWebSocket#291]
As seen in `IXWebSocketMessage.h`, the message class has a `const string&` member which binds to a reference to constant string  in the constructor. Since constant reference members don't prolong the lifetime of an object this can lead to undefined behavior, simply by doing 

```cpp
WebSocketMessage msg(type, "temporary string", ...);
// U.B when called with a temporary ^^^^
```

A description of the issue can be found [here](https://stackoverflow.com/q/2784262/4224575), while a canonical solution would be to use a string value in the constructor and move it into the member string.

I understand that (possibly) the way everything is designed there never is a temporary bound to the member variable. If that's the case, it would be nice to "safeguard" this assumption by deleting the respective r-value reference constructor like so: 

```
WebSocketMessage(..., std::string&&, ...) = delete;
```

Example [here](http://coliru.stacked-crooked.com/a/40cbd4ca4a17163b). Would an MR to add this check be a good idea for me to submit?

**Comment by bsergean** [comment]
You seem more knowledgeable than myself on this, so I'd be happy to look at a PR. The design goal as you probably guessed is to avoid expensive copy, so if you have a solution that is safer it would be great to use it.

A use case for the client is in this code, but I don't know how easy it would be to check that your change does not break that code. (it uses conan). Hopefully the unittest is sufficient.
https://github.com/machinezone/cobra/blob/master/sdk/cpp/ixcobra/ixcobra/IXCobraConnection.cpp

**Comment by picanumber** [comment]
TL;DR there's undefined behavior. Sorry for the long comment, but here's how to expose it. Declare the following constructor as deleted:

```cpp
WebSocketMessage(WebSocketMessageType t,
                              std::string&& s,
                              size_t w,
                              WebSocketErrorInfo e,
                              WebSocketOpenInfo o,
                              WebSocketCloseInfo c,
                              bool b = false) = delete;
```

i.e. refuse to construct the object when a temporary (`std::string&&`) is given to the constructor. The reason to do this is because the member `const std::string& str` [will not extend the lifetime of any temporaries passed to the constructor](https://stackoverflow.com/questions/2784262/does-a-const-reference-class-member-prolong-the-life-of-a-temporary). After doing this, compilation errors points to the point to where this happens, resulting in undefined behavior namely calls of `_onMessageCallback` with an empty string temporary (constructed by `""`) in `IXWebSocket.cpp`. 

In [this Pull Request](https://github.com/machinezone/IXWebSocket/pull/294) I suggest a fix for the current UB and place the deleted constructor to safeguard future misuses of the class. It seemed like the simplest solution given your requirement to never make copies.

**Comment by bsergean** [comment]
Thanks for the fix and the detailed explanation.
